src/approaches/grammar_search_invention_approach.py

```diff
@@ -820,14 +827,56 @@ def _evaluate_with_operators(self,
     @staticmethod
     def _get_refinement_prob(
             demo_atoms_sequence: Sequence[Set[GroundAtom]],
-            plan_atoms_sequence: Sequence[Collection[GroundAtom]]) -> float:
+            plan_atoms_sequence: Sequence[Set[GroundAtom]],
+            demo_multistep_effects: Set[FrozenSet[GroundAtom]]) -> float:
         """Estimate the probability that plan_atoms_sequence is refinable using
         the demonstration demo_atoms_sequence."""
-        # Make a soft assumption that the demonstrations are optimal.
+        # Make a soft assumption that the demonstrations are optimal,
+        # using a geometric distribution.
         demo_len = len(demo_atoms_sequence)
         plan_len = len(plan_atoms_sequence)
+        # The exponent is the difference in plan lengths.
+        exponent = abs(demo_len - plan_len)
+        if CFG.grammar_search_expected_nodes_include_suspicious_score:
+            # Handle suspicious effect scoring via unification.
+            num_suspicious_eff = 0
+            for i in range(len(plan_atoms_sequence) - 1):
+                for j in range(i + 1, len(plan_atoms_sequence)):
+                    atoms_i = plan_atoms_sequence[i]
+                    atoms_j = plan_atoms_sequence[j]
+                    for plan_eff in [
+                            frozenset(atoms_i - atoms_j),
+                            frozenset(atoms_j - atoms_i)
+                    ]:
+                        if not any(
+                                utils.unify(demo_eff, plan_eff)[0]
```

Comment:
how much slower does the overall score function get when using suspicious transitions?

Reply:
Speed:
without suspicious score:
* cover: 2.040 sec
* blocks: 21.633 sec
* painting: 25.072 sec

with suspicious score:
* cover: 2.341 sec
* blocks: 31.537 sec
* painting: 85.906 sec

So yeah as expected it's a lot slower when the demos are longer, but there's lots of things we can do to speed it up like only doing some multistep effects.

Reply:
thanks for checking that. what are those times exactly? hopefully not a single evaluation...

Reply:
the analysis pipeline which evaluates 7 pred sets for cover, 10 for blocks, 12 for painting

Reply:
so it's probably unusably slow in the current form for painting, but that's okay...